File: Leetcode/Python/_57.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        l = bisect.bisect_left([x[1] for x in intervals],newInterval[0])
        r = bisect.bisect_right([x[0] for x in intervals],newInterval[1])
        
        if l < len(intervals):
            newInterval[0] = min(newInterval[0],intervals[l][0])
            
        if r > 0:
            newInterval[1] = max(newInterval[1],intervals[r-1][1])
            
        return intervals[:l] + [ newInterval ] + intervals[r:]
      
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        s1, e1 = newInterval
        output = []
        index = 0
        for i, (s, e) in enumerate(intervals):
            if s < s1:
                output.append([s, e])
            else:
                index = i
                break

        if len(output) == 0 or output[-1][1] < s1:
            output.append([s1, e1])
        else:
            output[-1][1] = max(output[-1][1], e1)

        for s, e in intervals[index:]:
            if output[-1][1] < s:
                output.append([s, e])
            else:
                output[-1][1] = max(output[-1][1], e)

        return output

File: Leetcode/cpp/57.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        std::vector<std::vector<int>> output;
        int i = 0;
        for ( ; i < intervals.size(); ++i) {
            if (intervals[i][1] < newInterval[0]) {
                output.push_back(intervals[i]);
            } else {
                break;
            }
        }
        if (i == intervals.size()) {
            output.push_back(newInterval);
            return output;
        }
        if (intervals[i][0] > newInterval[0] && intervals[i][0] > newInterval[1]) {
            output.push_back(newInterval);
            output.push_back(intervals[i]);
        } else if (intervals[i][1] >= newInterval[0]) {
            output.push_back(
                {
                    std::min<int>(intervals[i][0], newInterval[0]),
                    std::max<int>(intervals[i][1], newInterval[1])
                }
            );
        }
        for (int j = i + 1; j < intervals.size(); ++j) {
            if (output.back()[1] >= intervals[j][0]) {
                output.back()[1] = std::max(output.back()[1], intervals[j][1]);
            } else {
                output.push_back(intervals[j]);
            }
        }
        return output;
    }
};
